bindings/pyroot/pythonizations/python/ROOT/_pythonization/_function_utils.py

```diff
@@ -0,0 +1,224 @@
+        else:
+            self.rdf = None
+            self.col_names = []
+    
+    def find_type(self, x, rdf = None):
+        """
+        Function to determine the type of a variable x.
+        If x is a string. It maps it to the corresponding column's type in RDF. (String constants are not supported)
+        Else it determines the fundemental type of x. (int, float, double)
+        Else it is an a numpy array and maps it to corresponding RVec.
+        Otherwise falgs a type error
+        Args:
+	    `   x: Variable whose type is to be determined
+
+        Returns:
+            type of x
+        """
+        try:
+            import numpy as np
+        except:
+            raise ImportError("Failed to import numpy during call to determine function signature.")
+        if isinstance(x, str):
+            #Can be string constant or can be column name
+            if x in self.col_names: # If x is a column
+                t = self.rdf.GetColumnType(x)
+                if t in map_tree_to_numbadeclares.keys(): # The column is a fundemental type from tree  
+                    return map_tree_to_numbadeclares[t]
+                elif '<' in  t: # The column type is a RVec<type>
+                    g = re.match('(.*)<(.*)>', t).groups(0)
+                    if g[1] in map_tree_to_numbadeclares.keys():
+                        return "RVec<" + map_tree_to_numbadeclares[g[1]] + ">"
+                    return "RVec<" + str(g[1]) + ">" # There are data type that leak into here. NOt sure from where. BUt need to implement something here.
+                    
+                else:
+                    return t
+            else :
+                return 'str'
+                #! Numba Declare does not support "string" type. Check _numbadeclare.Thus, Cannot pass string constants to the filter/Defines..
+        elif isinstance(x, bool):
+            return 'bool'
+        elif isinstance(x, int):
+            return 'int'
+        elif isinstance(x, float):
+            return 'float'
+        elif isinstance(x, np.ndarray): 
+            if x.dtype.type in  map_numpy_to_tree:
+                return "RVec<" + map_numpy_to_tree[x.dtype.type] + ">"
+            else:
+                raise TypeError(f"Support for {x.dtype.type} arrays is not yet supported.")
+        #! Need to work out how to map things like tuples, dicts, lists...
+        else:
+                raise TypeError(f"Type of {type(x)}:  {x} is not supported to be jitted.")
+            
+    def find_function_signature(self, func, funcargs):
+        """
+        Function to determine the function signature i.e the type of variables the function arguments and its corresponding value.
+        Args:
+	    `   func: Function whose signature is to be determined.
+                    P.S The function should not be crazy. It should have simple datatypes for the arguments and should not have default arguments. (Use funcargs instead)
```

Comment:
This first sentence ("crazy") is a bit colloquial :) I'd rephrase it.

Reply:
I think that any explanation about how the function that is passed to Filter should be, it would be better placed in the docstrings for the pythonization of Filter. The doxygen documentation in the RDF pythonizations file (at the very beginning of the file) should be updated with examples of the pythonization and what is supported -- we can delay this until we agree on the programming model, though.

Reply:
Yes. Sorry.